editor/src/clj/editor/dialogs.clj

```diff
@@ -660,19 +670,52 @@
          filter-term (or (:filter options)
                          (some-> filter-atom deref)
                          "")
-         initial-filtered-items (vec (filter-fn filter-term items))
+         filter-future-atom (atom nil)
+         state-atom (atom {:filter-term ""
+                           :filter-in-progress false
+                           :filtered-items []
+                           :selected-indices []})
+
+         set-filter-term
+         (fn set-filter-term [state filter-term]
+           (if (= (:filter-term state) filter-term)
+             state
+             (do (swap! filter-future-atom
+                        (fn [pending-filter-future]
+                          (thread-util/cancel-future! pending-filter-future)
+                          (future
+                            (try
+                              (let [filtered-items (vec (filter-fn filter-term items))
+                                    selected-indices (if (coll/empty? filtered-items) [] [0])]
+                                (thread-util/throw-if-interrupted!)
+                                (swap! state-atom
+                                       (fn [state]
+                                         (thread-util/throw-if-interrupted!)
+                                         (assoc state
+                                           :filter-in-progress false
+                                           :filtered-items filtered-items
+                                           :selected-indices selected-indices))))
+                              (catch InterruptedException _
+                                nil)
+                              (catch Throwable error
+                                (error-reporting/report-exception! error))))))
+                 (assoc state
+                   :filter-term filter-term
+                   :filter-in-progress true))))
+
+         _ (swap! state-atom set-filter-term filter-term)
+         event-handler (select-list-dialog-event-handler set-filter-term)
          result (fxui/show-dialog-and-await-result!
-                  :initial-state {:filter-term filter-term
-                                  :filtered-items initial-filtered-items
-                                  :selected-indices (if (seq initial-filtered-items) [0] [])}
-                  :event-handler (select-list-dialog-event-handler filter-fn items)
+                  :state-atom state-atom
+                  :event-handler event-handler
```

Comment:
The `:initial-state` has been moved into the `state-atom`, which we need access to from the outside so that the completed filter `future` can set the `:filtered-items` and so on when it completes.